src/sort.rs

```diff
@@ -0,0 +1,174 @@
+    ///         .children()
+    ///         .map(|n| n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    ///
+    /// // Example for sort_by_id.
+    /// tree.root_mut().sort_by(|a, b| a.id().cmp(&b.id()));
+    /// assert_eq!(
+    ///     vec![&'c', &'d', &'b'],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    /// ```
+    pub fn sort_by<F>(&mut self, mut compare: F)
+    where
+        F: FnMut(&NodeRef<T>, &NodeRef<T>) -> Ordering,
+    {
+        if self.has_children() {
+            let ids = self.sorted_ids(|nodes| {
+                nodes.sort_by(|a, b| compare(a, b));
+            });
+
+            for id in ids {
+                self.append_id(id);
+            }
+        }
+    }
+
+    /// Sort children by `NodeRef`'s key in ascending order using a key extraction function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ego_tree::tree;
+    ///
+    /// let mut tree = tree!("1a" => { "2b", "4c", "3d" });
+    /// tree.root_mut().sort_by_key(|a| a.value().split_at(1).0.parse::<i32>().unwrap());
+    /// assert_eq!(
+    ///     vec!["2b", "3d", "4c"],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| *n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    ///
+    /// // Example for sort_by_id.
+    /// tree.root_mut().sort_by_key(|n| n.id());
+    /// assert_eq!(
+    ///     vec![&"2b", &"4c", &"3d"],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    /// ```
+    pub fn sort_by_key<K, F>(&mut self, mut f: F)
+    where
+        F: FnMut(&NodeRef<T>) -> K,
+        K: Ord,
+    {
+        if self.has_children() {
+            let ids = self.sorted_ids(|nodes| {
+                nodes.sort_by_key(|node| f(node));
+            });
+
+            for id in ids {
+                self.append_id(id);
+            }
+        }
+    }
+
+    /// Returns a vector of `NodeId`s of the node's children, sorted according to
+    /// the provided closure.
+    ///
+    /// The closure `f` is applied to a mutable vector of `NodeRef`s representing
+    /// the children of the node. This allows you to define the sorting logic
+    /// within the closure, which will then determine the order of the returned
+    /// `NodeId`s.
+    ///
+    /// # Parameters
+    ///
+    /// - `f`: A closure that takes a mutable reference to a vector of `NodeRef`s.
+    ///         The closure is responsible for sorting the vector in-place.
+    ///
+    /// # Returns
+    ///
+    /// A vector of `NodeId`s representing the children of the node, sorted
+    /// according to the logic defined in the closure `f`.
+    pub fn sorted_ids<F>(&self, mut f: F) -> Vec<NodeId>
+    where
+        F: FnMut(&mut Vec<NodeRef<T>>),
+    {
+        let node_ref = unsafe { self.tree.get_unchecked(self.id) };

```

Comment:
`NodeRef` cannot from `NodeMut` without transferring ownership, so `NodeMut` also implements this method use unsafe.

Reply:
The signature is unsound as it exposes read-only node wrappers from a mutable one. For example, one could use the returned `Vec` and use its `parent` methods to get an aliasing reference to `self`.

Reply:
Maybe `F: FnMut(&mut Vec<(NodeId, &T)>)` is good. Because `F: FnMut(&mut Vec<NodeMut<T>>)` but cannot create a new `Vec<NodeMut<T>>`.

Reply:
Yes, especially since `sort_by` is unsound for the same reason, i.e. we can only expose `(NodeId, &T)` here.

But as written below, this is all not something I am keen to expose in an API the implementation of which we might reconsider eventually.

Reply:
> Yes, especially since sort_by is unsound for the same reason, i.e. we can only expose (NodeId, &T) here.

Actually no, that was wrong. We can expose `NodeRef` via methods that take `&mut self` thereby "shadowing" the unique reference behind `NodeMut/self`.

Reply:
Please also just fold this method body into `sort_by`, which is the only place it needs to be used. KISS applies, i.e. let's made this the simplest and most obviously correct version for this PR.

Reply:
Folding it also allows us to avoid `collect`ing twice which is probably more expensive than the additional `append_id` calls.

Reply:
> Updated (NodeId, &T) version.

As written above, we can use the three methods and expose `NodeRef` inside, we just have to make sure the method taking the callback uses `&mut self` and thereby exclusively owns the `NodeMut` (and thereby the whole `Tree`).

Reply:
`F: FnMut(&NodeMut<T>, &NodeMut<T>) -> Ordering`

```rust
        nodes.sort_by(|a, b| {
            compare(
                unsafe { &self.tree.get_unchecked_mut(a.0) },
                unsafe { &self.tree.get_unchecked_mut(b.0) },
            )
```

cannot borrow `**self.tree` as mutable more than once at a time second mutable borrow occurs here

Reply:
> Folding it also allows us to avoid `collect`ing twice which is probably more expensive than the additional `append_id` calls.

Twice `collect` is necessary, `append_id` cannot in closure.

Reply:
> > Yes, especially since sort_by is unsound for the same reason, i.e. we can only expose (NodeId, &T) here.
> 
> Actually no, that was wrong. We can expose `NodeRef` via methods that take `&mut self` thereby "shadowing" the unique reference behind `NodeMut/self`.

Sorry, I didn't see this message. Too much information seems to be inconvenient to use.

Reply:
```rust
pub fn sort_by<F>(&mut self, mut compare: F)
where
    F: FnMut(NodeRef<T>, NodeRef<T>) -> Ordering,
{
    if !self.has_children() {
        return;
    }

    let mut children = {
        let this = unsafe { self.tree.get_unchecked(self.id) };
        this.children().map(|child| child.id).collect::<Vec<_>>()
    };

    children.sort_by(|a, b| {
        let a = unsafe { self.tree.get_unchecked(*a) };
        let b = unsafe { self.tree.get_unchecked(*b) };
        compare(a, b)
    });

    for id in children {
        self.append_id(id);
    }
}
```

Reply:
`F: FnMut(NodeRef<T>, NodeRef<T>) -> Ordering` vs `F: FnMut(&NodeRef<T>, &NodeRef<T>) -> Ordering`. 
Is there a clear difference between these two?

Reply:
```rust
        let a = unsafe { self.tree.get_unchecked(*a) };
        let b = unsafe { self.tree.get_unchecked(*b) };
```
It seems that each id `get_unexecked` multiple times in `sort_by`. But it's okay, it reduced the vec allocation space once.

Reply:
> Is there a clear difference between these two?

Yes. `NodeRef` is already a reference so `&NodeRef` is an unnecessary indirection and passing `NodeRef` by value should be preferred. (SROA will optimize all of this away in any case.)

> It seems that each id get_unchecked multiple times in sort_by. 

This does not cost anything as it is the unsafe version without bounds checking.